### PR TITLE
Add `util.inspect` emulation in REPL

### PR DESCRIPTION
**Pull request by chqrlie**
- output values with controlable depth and detail
- add `.hidden` and `.depth` directives
- remove `eval_mode`
- add `use_strict` and `.strict` meta command
- add missing closures on global objects
- save and load command history to/from `~/.qjs_history`

### CODE REVIEW

File: repl.js [hunk 1]
```diff
@@ -1396,15 +1792,39 @@ import * as os from "os";
...
+        if (s) {
+            try {
+                var f = std.open(config_file(".qjs_history"), "w");
+                f.puts(s + '\n');
+                f.close();
+            } catch (e) {
+            }
+        }
+    }
+    function load_history() {
```
Comment (saghul) [inline repl.js:1833]: Oh, nice one!

File: repl.js [hunk 2]
```diff
@@ -945,84 +966,451 @@ import * as os from "os";
...
-    function print(a) {
-        var stack = [];
-
-        function print_rec(a) {
-            var n, i, keys, key, type, s;
-            
-            type = typeof(a);
-            if (type === "object") {
+
+    var util = {};
```
Comment (saghul) [inline repl.js:975]: What's the purpose of having the `util` module? We could just have the `inspect` function here and comment that it mimics Node's util.inspect?

File: repl.js [hunk 3]
```diff
@@ -945,84 +966,451 @@ import * as os from "os";
...
+        }
+        function push_token(s) {
+            tokens.push("" + s);
+        }
+        function append_token(s) {
+            tokens[tokens.length - 1] += s;
+        }
+        function class_tag(o) {
+            /* return the class id of an object */
+            return Object.prototype.toString.call(o).slice(8, -1);
```
Comment (saghul) [inline repl.js:1016]: Maybe get the strint tag symbol instead?

File: repl.js [hunk 4]
```diff
@@ -945,84 +966,451 @@ import * as os from "os";
...
+        }
+
+        function print_rec(a, level) {
+            var n, n0, i, k, keys, key, type, isarray, noindex, nokeys, brace, sep;
+
+            switch (type = typeof(a)) {
+            case 'undefined':
```
Comment (saghul) [inline repl.js]: Minor style nit: the other strings are quoted with double quotes...

File: repl.js [hunk 5]
```diff
@@ -1396,15 +1792,39 @@ import * as os from "os";
         return [ state, level, r ];
     }
 
-    var m, s = std.getenv("COLORFGBG");
-    if (s && (m = s.match(/(\d+);(\d+)/))) {
-        if (+m[2] !== 0) { // light background
-            styles = themes.light;
+    function config_file(s) {
+        return (std.getenv("HOME") || ".") + "/" + s;
```
Comment (saghul) [inline repl.js]: I don't think HOME will work on Windows. The path separator is also `\\`

File: repl.js [hunk 6]
```diff
@@ -1396,15 +1792,39 @@ import * as os from "os";
...
+    }
+    function save_history() {
+        var s = history.slice(-1000).join('\n').trim();
```
Comment (saghul) [inline repl.js:1823]: Any reason not to save it as JSON?

**Review by saghul — COMMENTED** [review]
Some great improvements here!

File: repl.js (hunk 2 above)
Comment (chqrlie) [reply repl.js:975]: I used this as a temporary solution, exposing the `util` module to the user would be nice too.

File: repl.js (hunk 3 above)
Comment (chqrlie) [reply repl.js:1016]: You mean the `Symbol.toStringTag` property? Yes that would be cleaner.

File: repl.js (hunk 5 above)
Comment (chqrlie) [reply repl.js]: The `/` path separator should work on Windows too.  How can we get the user home dir on Windows?

File: repl.js (hunk 6 above)
Comment (chqrlie) [reply repl.js:1823]: A simple text file is much simpler and more standard.  The only limitation is we cannot save multiline commands, which we don't have at the moment.

File: repl.js (hunk 3 above)
Comment (saghul) [reply repl.js:1016]: YEah that one. The only caveat is not every object has it...

File: repl.js (hunk 2 above)
Comment (saghul) [reply repl.js:975]: 👍

File: repl.js (hunk 5 above)
Comment (saghul) [reply repl.js]: https://github.com/libuv/libuv/blob/8083ab26e0f7c163d131a0509f876f50ed4c110d/src/win/util.c#L934

We could port that and the unix equivalent, which takes care of many corner cases, as `std.homedir` perhaps.

File: repl.js (hunk 6 above)
Comment (saghul) [reply repl.js:1823]: 👍

File: repl.js (hunk 5 above)
Comment (chqrlie) [reply repl.js]: It seems `getenv("USERPROFILE")` should do the job.  I will change to:
```
    function config_file(s) {
        return (std.getenv("HOME") || std.getenv("USERPROFILE") || ".") + "/" + s;
    }
```

File: repl.js (hunk 3 above)
Comment (chqrlie) [reply repl.js:1016]: Many objects do not have this property and for those who do the current method works.
It also works for for boxed objects, `Math`, `JSON`, `globalThis`, etc. So we should keep this convoluted but reliable method.

Comment (saghul) [reply repl.js:1016]: Makes sense, thanks for checking though!

File: repl.js (hunk 5 above)
Comment (saghul) [reply repl.js]: Sounds good. We can add some extra functionality to std at a later stage.

File: repl.js (hunk 4 above)
Comment (chqrlie) [reply repl.js]: fixed

**Review by saghul — APPROVED** [review]

**Review by saghul — APPROVED** [review]

**Pull request merged**